cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -0,0 +1,190 @@
+ * the License.
+ */
+
+package io.cdap.cdap.internal.tethering;
+
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.internal.remote.RemoteAuthenticator;
+import io.cdap.cdap.common.service.AbstractRetryableScheduledService;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.internal.app.store.AppMetadataStore;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+/**
+ * The main class to run the remote agent service.
+ */
+public class TetheringAgentService extends AbstractRetryableScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(TetheringAgentService.class);
+  private static final Gson GSON = new Gson();
+  private static final String CONNECT_CONTROL_CHANNEL = "/v3/tethering/controlchannels/";
+  private static final String SUBSCRIBER = "tether.agent";
+  private static final String TOPIC = "tethering";
+
+  private final CConfiguration cConf;
+  private final long connectionInterval;
+  private final TetheringStore store;
+  private final String instanceName;
+  private final TransactionRunner transactionRunner;
+  private String lastMessageId;
+
+  @Inject
+  TetheringAgentService(CConfiguration cConf, TransactionRunner transactionRunner) {
+    super(RetryStrategies.fixDelay(cConf.getLong(Constants.Tethering.CONNECT_INTERVAL), TimeUnit.SECONDS));
+    this.connectionInterval = TimeUnit.SECONDS.toMillis(cConf.getLong(Constants.Tethering.CONNECT_INTERVAL));
+    this.cConf = cConf;
+    this.transactionRunner = transactionRunner;
+    this.store = new TetheringStore(transactionRunner);
+    this.instanceName = cConf.get(Constants.INSTANCE_NAME);
+  }
+
+  @Override
+  protected void doStartUp() throws InstantiationException, IllegalAccessException {
+    Class<? extends RemoteAuthenticator> authClass = cConf.getClass(Constants.Tethering.CLIENT_AUTHENTICATOR_CLASS,
+                                                                    null,
+                                                                    RemoteAuthenticator.class);
+    if (authClass != null) {
+      RemoteAuthenticator.setDefaultAuthenticator(authClass.newInstance());
+    }
+    TransactionRunners.run(transactionRunner, context -> {
+      lastMessageId = AppMetadataStore.create(context).retrieveSubscriberState(TOPIC, SUBSCRIBER);
```

Comment:
Why there is only one messageId? What happen if it connects to multiple CDAP?

Reply:
oh right, thanks for pointing that out. changed to map of message ids to handle multiple peers